### PR TITLE
codeintel: Support bulk deletion of LSIF uploads

### LINKED ISSUE #41193: codeintel: Support bulk operations
Inspired by #40708. We'd like to be able to do some bulk operations on lsif_upload/index records. For both types of records, we'd benefit from having a bulk deletion action. For index records, we'd benefit from being able to mark a selection of indexes as "re-indexable" en masse so that we can ship indexer updates without having customers lose intelligence.

This ticket covers the UI additions to the bulk operation feature, allowing a user to select individual records in the list view, or allowing users to select all records that _match_ the given list view query and send some sort of bulk response to the backend.

Tree of issues:

- Original motivation https://github.com/sourcegraph/sourcegraph/issues/40708
  - LSIF uploads: implementing in https://github.com/sourcegraph/sourcegraph/pull/42395
    - #41194
  - LSIF indexes: implementing in TODO
    - #41196
    - #41195

Requested by:

- https://github.com/sourcegraph/accounts/issues/8205
- https://github.com/sourcegraph/accounts/issues/6036

### PR DESCRIPTION
**Pull request by chrismwendt**
Resolves #41193 and #41194.

CleanShot 2022-09-30 at 16 56 52@2x

## Test plan

Upload some LSIF dumps, visit http://localhost:3080/site-admin/code-graph/uploads, check/uncheck, delete, select matching, delete matching.

### CODE REVIEW

File: internal/database/migration/shared/upgradedata/stitched-migration-graph.json [hunk 1]
```diff
@@ -6742,10 +6742,7 @@
 			"v3.42.0": {
 				"RootID": 1646027072,
 				"LeafIDs": [
-					1654770608,
-					1658174103,
-					1658225452,
-					1657663493
+					1657635365
```
Comment (efritz) [inline internal/database/migration/shared/upgradedata/stitched-migration-graph.json:6745]: 🤔 

Seems like you have some lint in CI as well (and need to regenerate mocks). You should run `go generate ./...` in the root to fix these errors.

**Review by efritz — CHANGES_REQUESTED** [review]

File: internal/codeintel/uploads/internal/store/store_uploads.go [hunk 2]
```diff
@@ -675,6 +675,59 @@ const deleteUploadByIDQuery = `
 UPDATE lsif_uploads u SET state = CASE WHEN u.state = 'completed' THEN 'deleting' ELSE 'deleted' END WHERE id = %s RETURNING repository_id
 `
 
+// DeleteUploads deletes uploads by filter criteria. The associated repositories will be marked as dirty
+// so that their commit graphs will be updated in the background.
+func (s *store) DeleteUploads(ctx context.Context, opts types.DeleteUploadsOptions) (err error) {
```
Comment (efritz) [inline internal/codeintel/uploads/internal/store/store_uploads.go:680]: Can we get a unit test for this?

File: internal/codeintel/uploads/internal/store/store_uploads.go [hunk 3]
```diff
@@ -675,6 +675,59 @@ const deleteUploadByIDQuery = `
...
+	conds = append(conds, authzConds)
+
+	tx, err := s.transact(ctx)
+	if err != nil {
+		return err
+	}
+	defer func() { err = tx.Done(err) }()
+
+	unset, _ := tx.db.SetLocal(ctx, "codeintel.lsif_uploads_audit.reason", "direct delete by filter criteria request")
+	defer unset(ctx)
```
Comment (efritz) [inline internal/codeintel/uploads/internal/store/store_uploads.go:698]: <3

File: client/web/src/enterprise/codeintel/uploads/pages/CodeIntelUploadsPage.tsx [hunk 4]
```diff
@@ -164,6 +191,61 @@ export const CodeIntelUploadsPage: FunctionComponent<React.PropsWithChildren<Cod
...
+
+                                return
+                            }
+
+                            for (const id of selection) {
+                                await handleDeleteLsifUpload({
+                                    variables: { id },
+                                    update: cache => cache.modify({ fields: { node: () => {} } }),
+                                })
+                            }
```
Comment (efritz) [inline client/web/src/enterprise/codeintel/uploads/pages/CodeIntelUploadsPage.tsx:231]: Would it be good idea (follow-up PR would be good too) to batch these into a single request?

Comment (chrismwendt) [reply client/web/src/enterprise/codeintel/uploads/pages/CodeIntelUploadsPage.tsx:231]: Chatted in Zoom: it should be fast enough as-is, and we can batch it in a separate PR later if anyone notices it being slow.

**Review by efritz — APPROVED** [review]

**Pull request merged**